File: test-suite/src/test/java/io/micronaut/docs/validation/iterable/BookInfoSpec.java
package io.micronaut.docs.validation.iterable;

import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
import jakarta.inject.Inject;
import jakarta.validation.ConstraintViolationException;
import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

@MicronautTest
public class BookInfoSpec {

    @Inject
    BookInfoService bookInfoService;

    // tag::validate-iterables[]

    @Test
    void testAuthorNamesAreValidated() {
        final List<String> authors = Arrays.asList("Me", "");

        final ConstraintViolationException exception =
                assertThrows(ConstraintViolationException.class, () ->
                        bookInfoService.setBookAuthors("My Book", authors)
                );

        assertEquals("setBookAuthors.authors[1]<list element>: must not be blank",
                exception.getMessage()); // <1>
    }

    @Test
    void testSectionsAreValidated() {
        final Map<String, Integer> sectionStartPages = new HashMap<>();
        sectionStartPages.put("", 1);

        final ConstraintViolationException exception =
                assertThrows(ConstraintViolationException.class, () ->
                        bookInfoService.setBookSectionPages("My Book", sectionStartPages)
                );

        assertEquals("setBookSectionPages.sectionStartPages[]<map key>: must not be blank",
                exception.getMessage()); // <2>
    }

    // end::validate-iterables[]
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/docs/validation/iterable/BookInfoSpec.kt
package io.micronaut.docs.validation.iterable

import io.micronaut.test.extensions.junit5.annotation.MicronautTest
import jakarta.inject.Inject
import jakarta.validation.ConstraintViolationException
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

@MicronautTest
class BookInfoSpec {
    @Inject
    lateinit var bookInfoService: BookInfoService

    // tag::validate-iterables[]
    @Test
    fun testAuthorNamesAreValidated() {
        val authors: List<String> = mutableListOf("Me", "")
        val exception = Assertions.assertThrows(
            ConstraintViolationException::class.java
        ) { bookInfoService.setBookAuthors("My Book", authors) }
        Assertions.assertEquals(
            "setBookAuthors.authors[1]<list element>: must not be blank",
            exception.message
        ) // <1>
    }

    @Test
    fun testSectionsAreValidated() {
        val sectionStartPages: MutableMap<String, Int> = HashMap()
        sectionStartPages[""] = 1
        val exception = Assertions.assertThrows(
            ConstraintViolationException::class.java
        ) {
            bookInfoService.setBookSectionPages(
                "My Book",
                sectionStartPages
            )
        }
        Assertions.assertEquals(
            "setBookSectionPages.sectionStartPages[]<map key>: must not be blank",
            exception.message
        ) // <2>
    } // end::validate-iterables[]
}
